File: c-sharp/src/1.SinglyLinkedList.cs
using System;
using System.Text;

namespace DataStructures.LinkedList {
    public class Node<T> {
        public T Value { get; set; }
        public Node<T> Next { get; set; }

        public Node (T value) {
            this.Value = value;
            this.Next = null;
        }
    }

    public class SinglyLinkedList<T> {
        #region Properties

        public int Length { get; set; }

        public Node<T> Head { get; set; }

        public Node<T> Tail { get; set; }

        #endregion

        public SinglyLinkedList () {
            this.Length = 0;
            this.Head = null;
            this.Tail = null;
        }

        /// <summary>
        /// Adds a node at the end of the linked list
        /// </summary>
        /// <param name="value">Value for the new node that needs to be added the linked list</param>
        public bool Push (T value) {
            Node<T> newNode = new Node<T> (value);

            if (this.Length == 0) {
                this.Head = newNode;
                this.Tail = newNode;
            } else {
                this.Tail.Next = newNode;
                this.Tail = newNode;
            }
            this.Length++;
            return true;
        }

        /// <summary>
        /// Removes a node at the end of the linked list and returns the node
        /// </summary>
        public Node<T> Pop () {
            if (this.Length == 0) {
                return null;
            } else if (this.Head == this.Tail) {
                var temp = this.Head;
                this.Head = null;
                this.Tail = null;
                this.Length--;

                return temp;
            } else {
                // Find the node just before the tail node
                Node<T> prevNode = null;
                Node<T> currentNode = this.Head;
                while (currentNode.Next != null) {
                    prevNode = currentNode;
                    currentNode = currentNode.Next;
                }
                // Set the tail to the previous node
                var temp = this.Tail;
                this.Tail = prevNode;
                this.Tail.Next = null;
                this.Length--;

                return temp;
            }
        }

        /// <summary>
        /// Removes a node from the beginning of the list and returns the node.
        /// </summary>
        public Node<T> Shift () {
            if (this.Length == 0) return null;
            else if (this.Head == this.Tail) {
                var temp = this.Head;
                this.Head = null;
                this.Tail = null;
                this.Length = 0;

                return temp;
            } else {
                var temp = this.Head;
                this.Head = temp.Next;
                temp.Next = null;
                this.Length--;

                return temp;
            }
        }

        /// <summary>
        /// Adds a node at the begining of the linked list
        /// </summary>
        /// <param name="value">Value for the new node that needs to be added the linked list</param>
        public bool UnShift (T value) {
            Node<T> newNode = new Node<T> (value);
            if (this.Length == 0) {
                this.Head = newNode;
                this.Tail = newNode;
            } else {
                newNode.Next = this.Head;
                this.Head = newNode;
            }
            this.Length++;
            return true;
        }

        ///<summary>
        ///Returns the node at the specified index of the linked list.
        ///If index = 0; first element in the list is returned.
        ///If index = 3; fourth element in the list is returned.
        /// </summary>
        /// <param name="index">Index of the node to be retrieved</param>
        public Node<T> Get (int index) {
            if (this.Length == 0 || index < 0 || index > this.Length - 1) return null;

            var currentNode = this.Head;
            for (int i = 1; i <= index; i++) {
                currentNode = currentNode.Next;
            }
            return currentNode;
        }

        /// <summary>
        /// Updates the value of the node at the specified index.
        /// If index = 0; first element in the list is returned.
        /// If index = 3; fourth element in the list is returned.
        /// </summary>
        /// <param name="index">Index of the node to be updated</param>
        /// <param name="newValue">New value for the node to be updated</param>
        public bool Set (int index, T newValue) {
            var node = this.Get (index);
            if (node == null) {
                return false;
            } else {
                node.Value = newValue;
                return true;
            }
        }

        /// <summary>
        /// Inserts a new node at the specified index.
        /// </summary>
        /// <param name="index">Index of the node to be inserted</param>
        /// <param name="value">Value for the node to be inserted</param>
        public bool Insert (int index, T value) {
            if (index < 0 || index > this.Length) return false;

            if (index == 0) {
                this.UnShift (value);
                return true;
            } else if (index == this.Length) {
                this.Push (value);
                return true;
            } else {
                var newNode = new Node<T> (value);
                var pervNode = this.Get (index - 1);

                if (pervNode == null) {
                    return false;
                } else {
                    var nextNode = pervNode.Next;
                    pervNode.Next = newNode;
                    newNode.Next = nextNode;
                    this.Length++;

                    return true;
                }
            }
        }

        /// <summary>
        /// Removes a node at the specified index and returns the node.
        /// </summary>
        /// <param name="index">Index of the node to be removed</param>
        public Node<T> Remove (int index) {
            if (index < 0 || index > this.Length - 1) return null;
            if (index == 0) {
                return this.Shift ();
            } else if (index == this.Length - 1) {
                return this.Pop ();
            } else {
                var prevNode = this.Get (index - 1);
                var removedNode = prevNode.Next;
                prevNode.Next = removedNode.Next;
                removedNode.Next = null;
                this.Length--;

                return removedNode;
            }
        }

        /// <summary>
        /// Reverses the linked list. 1 -> 2 -> 3 -> 4
        /// </summary>
        public void Reverse () {
            if (this.Length != 0) {
                Node<T> prevNode = null;
                // Swap Head and Tail
                Node<T> currentNode = this.Head;
                this.Head = this.Tail;
                this.Tail = currentNode;

                // Reverse the list
                for (int i = 0; i < this.Length; i++) {
                    var nextNode = currentNode.Next;
                    currentNode.Next = prevNode;

                    prevNode = currentNode;
                    currentNode = nextNode;
                }
            }
        }

        public void Print () {
            StringBuilder data = new StringBuilder ();
            var currentNode = this.Head;
            while (currentNode != null) {
                data.Append (currentNode.Value + "   ");
                currentNode = currentNode.Next;
            }
            Console.WriteLine (data.ToString ());
        }
    }
}

File: js/src/1.singlyLinkedList.ts
class ListNode<T> {
  private _value: T;
  public get value(): T {
    return this._value;
  }
  public set value(v: T) {
    this._value = v;
  }

  private _next: ListNode<T>;
  public get next(): ListNode<T> {
    return this._next;
  }
  public set next(v: ListNode<T>) {
    this._next = v;
  }

  constructor(nodeValue: T) {
    this._value = nodeValue;
    this._next = null;
  }
}

class SinglyLinkedList<T> {
  //#region " Properties "

  private _head: ListNode<T>;
  public get head(): ListNode<T> {
    return this._head;
  }
  public set head(node: ListNode<T>) {
    this._head = node;
  }

  private _tail: ListNode<T>;
  public get tail(): ListNode<T> {
    return this._tail;
  }
  public set tail(node: ListNode<T>) {
    this._tail = node;
  }

  private _length: number;
  public get length(): number {
    return this._length;
  }
  public set length(v: number) {
    this._length = v;
  }

  //#endregion

  constructor() {
    this._length = 0;
    this._head = null;
    this._tail = null;
  }

  /**
   * Adds a node to the end of the linked list
   * @param value Value to be stored in the linked list node
   */
  push(value: T) {
    let newNode = new ListNode(value);

    // If head is null (empty linked list), then assign both head and tail to the new node
    if (this._head === null) {
      this._head = newNode;
      this._tail = this._head;
    } else {
      this._tail.next = newNode;
      this._tail = newNode;
    }
    // Increment the length of the linked list
    this._length++;
  }

  /**
   * Removes a node at the end of the linked list and will return the node value
   */
  pop(): T {
    if (this._length === 0) {
      return undefined;
    } else if (this._head === this._tail) {
      const returnNode: ListNode<T> = this._head;
      this._head = null;
      this._tail = null;
      this._length = 0;
      return returnNode.value;
    } else {
      let returnNode: ListNode<T>;
      let newTailNode: ListNode<T>;
      let currentNode = this._head;
      // Find the node before the tail node
      while (currentNode.next) {
        newTailNode = currentNode;
        currentNode = currentNode.next;
      }
      // Set the tail to the new tail node
      returnNode = this._tail;
      newTailNode.next = null;
      this._tail = newTailNode;
      // Decrement the length of the linked list
      this._length--;

      return returnNode.value;
    }
  }

  /**
   * Removes a node form the beginning of the linked list and will return the node value
   */
  shift(): T {
    if (this._length === 0) {
      return undefined;
    } else if (this._head === this._tail) {
      const returnNode = this._head;
      this._head = null;
      this._tail = null;
      this._length = 0;
      return returnNode.value;
    } else {
      const returnNode = this._head;
      this._head = returnNode.next;
      returnNode.next = null;
      this._length--;

      return returnNode.value;
    }
  }

  /**
   * Adds a node at the beginning of the linked list
   * @param value Value to be stored at the beginning of the linked list
   */
  unShift(value: T) {
    const node = new ListNode<T>(value);
    if (this._length === 0) {
      this._head = node;
      this.tail = node;
    } else {
      node.next = this._head;
      this._head = node;
    }
    this._length++;
  }

  /**
   * Returns the node at the specified index of the linked list.
   * If index = 0; first element in the list is returned.
   * If index = 3; fourth element in the list is returned.
   * @param index Index of the node to be retrieved
   */
  get(index: number): ListNode<T> {
    if (this._length === 0 || index > this._length - 1 || index < 0) {
      return undefined;
    } else {
      let returnNode: ListNode<T> = this._head;
      for (let i = 1; i <= index; i++) {
        returnNode = returnNode.next;
      }
      return returnNode;
    }
  }

  /**
   * Updates the value of the node at the specified index.
   * If index = 0; Value of the first element in the list is updated.
   * If index = 3; Value of the fourth element in the list is updated.
   * @param index Index of the node to be updated
   * @param value New value of the node
   */
  set(index: number, value: T): boolean {
    let node = this.get(index);
    if (node) {
      node.value = value;
      return true;
    } else {
      return false;
    }
  }

  /**
   * Inserts a new node at the specified index.
   * @param index Index at which the new node has to be inserted
   * @param value Value of the new node to be inserted
   */
  insert(index: number, value: T): boolean {
    if (index < 0 || index > this._length) {
      return false;
    } else if (index === 0) {
      this.unShift(value);
      return true;
    } else if (index === this._length) {
      this.push(value);
      return true;
    } else {
      const prevNode = this.get(index - 1);
      const nextNode = prevNode.next;

      const newNode = new ListNode<T>(value);
      prevNode.next = newNode;
      newNode.next = nextNode;
      this._length++;
      return true;
    }
  }

  /**
   * Removes a node at the specified index and returns its value.
   * @param index Index at which the node has to be removed.
   */
  remove(index: number): T {
    if (index < 0 || index > this._length - 1) {
      return null;
    } else if (index === 0) {
      return this.shift();
    } else if (index === this._length - 1) {
      return this.pop();
    } else {
      const prevNode = this.get(index - 1);
      const removeNode = prevNode.next;
      prevNode.next = removeNode.next;
      removeNode.next = null;
      this._length--;
      return removeNode.value;
    }
  }

  reverse() {
    //Swap head and tail nodes
    let node = this._head;
    this._head = this._tail;
    this._tail = node;

    let prevNode: ListNode<T> = null;
    let nextNode: ListNode<T>;
    for (let i = 0; i < this._length; i++) {
      nextNode = node.next;
      node.next = prevNode;
      prevNode = node;
      node = nextNode;
    }
  }

  print() {
    let arr: T[] = [];
    let current = this._head;
    while (current) {
      arr.push(current.value);
      current = current.next;
    }
    console.log(arr);
  }
}

let ls = new SinglyLinkedList<number>();
console.log(ls.length);
ls.push(0);
ls.push(1);
ls.push(2);

ls.print();

// console.log(ls.pop());
// console.log(ls.pop());
// console.log(ls.pop());
// console.log(ls.pop());

// console.log(ls.shift());
// console.log(ls.shift());
// console.log(ls.shift());
// console.log(ls.shift());

// console.log(ls.unShift(-1));
// console.log(ls.unShift(-2));
// console.log(ls.unShift(-3));
// console.log(ls.unShift(-4));

// console.log(ls.set(3, 100));
// console.log(ls.get(1));
// console.log(ls.insert(1, -1));
// console.log(ls.get(1));
ls.reverse();
ls.print();
